### PR TITLE
Remove 0.5 deprecations

### PR DESCRIPTION
**Pull request by StefanKarpinski**

### CODE REVIEW

**Comment by tkelman** [comment]
We need to go through and make sure they're all removed from the docs, and any comments left in deprecated.jl have been addressed. Also any syntax or C-level deprecations that were added in 0.5, but those can be a separate PR.

**Comment by StefanKarpinski** [comment]
There was only one TODO in the deprecations section, which was to remove the `isequal` methods for `Char` and `Integer` (done). @JeffBezanson has a [separate PR](https://github.com/JuliaLang/julia/pull/20327) for the syntax deprecations.

**Comment by tkelman** [comment]
ok. should git grep all the removals and clear them out from docs and anywhere else

**Comment by tkelman** [comment]
and there are things like getfield with parens that I think were done in C

**Comment by StefanKarpinski** [comment]
Looks like https://github.com/JuliaLang/julia/commit/7711aef11c755a08fde53423bfb6093203d5eb1d accidentally exported `trailingsize` even though it was never meant to be public. Deleting the 0.5 deprecations accidentally removed the implicit export of it. Not sure how to best handle that situation.

**Comment by StefanKarpinski** [comment]
I cherry-picked @JeffBezanson's syntax deprecations removal and looked through the C code for anything that's deprecated entirely there, but as far as I can tell it's only `invoke` with a tuple of types instead of a tuple type, `isdefined` for arrays, and partial linear indexing, all of which are new to 0.6. They should probably have "DEPRECATION 0.6" notes on them, however.

**Comment by StefanKarpinski** [comment]
Mysteriously, this branch has this strange situation:

```jl
julia> :String in names(Base, false, true)
false

julia> :String in names(Core, false, true)
true

julia> :Array in names(Base, false, true)
true

julia> :Array in names(Core, false, true)
true
```

**Comment by StefanKarpinski** [comment]
Mystery solved: we were not explicitly exporting `String` from `Base` before, it was incidentally exported via a deprecation. I'm not sure how undeprecating it didn't break everything, but hey.

File: base/deprecated.jl [hunk 1]
```diff
@@ -105,651 +105,7 @@ macro deprecate_binding(old, new)
...
-    if isdefined(Base, :STDOUT)
-        return displaysize(STDOUT)
-    end
-    return displaysize()
-end
-
-#14335
-@deprecate super(T::DataType) supertype(T)
-
-function with_output_limit(thk, lim::Bool=true) # thk is usually show()
```
Comment (tkelman) [inline base/deprecated.jl]: found a todo here, will push to this branch if you don't mind

File: base/deprecated.jl [hunk 2]
```diff
@@ -105,651 +105,7 @@ macro deprecate_binding(old, new)
...
-        :(@inbounds $(esc(blk)))
-    end
-end
-
-
-@deprecate parseip(str::AbstractString) parse(IPAddr, str)
-
-#https://github.com/JuliaLang/julia/issues/14608
-@deprecate readall readstring
-@deprecate readbytes read
```
Comment (tkelman) [inline base/deprecated.jl]: we still have, and export, `readbytes!`, ah well

File: base/deprecated.jl (hunk 1 above)
Comment (StefanKarpinski) [inline base/deprecated.jl]: What's the todo? I don't see it here.

File: base/deprecated.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/deprecated.jl]: yes, it's all over the place and I think should be?

File: base/deprecated.jl (hunk 1 above)
Comment (tkelman) [inline base/deprecated.jl]: just removing one line

**Comment by JeffBezanson** [comment]
Base doesn't need to export `String`, since `Core` exports it.

File: base/show.jl [hunk 3]
```diff
@@ -82,17 +82,10 @@ haskey(io::IOContext, key) = haskey(io.dict, key)
 haskey(io::IO, key) = false
 getindex(io::IOContext, key) = getindex(io.dict, key)
 getindex(io::IO, key) = throw(KeyError(key))
-_limit_output = nothing # TODO: delete with with_output_limit deprecation
```
Comment (StefanKarpinski) [inline base/show.jl]: These TODOs really need to go in `base/deprecated.jl` in the future. Thanks for finding this one.

**Comment by StefanKarpinski** [comment]
> Base doesn't need to export String, since Core exports it.

And yet, without that export, we have this on this branch:

```jl
julia> :String in names(Base, false, true)
false
```

So there seems to be a bug in `names` which exporting `String` from `Base` works around.

**Comment by StefanKarpinski** [comment]
Or perhaps I'm misunderstanding what that argument to `names` is supposed to do since there's a pretty large list of these:

```jl
julia> setdiff(names(Core), names(Base, false, true))
96-element Array{Symbol,1}:
 :(===)
 :ANY
 :AbstractArray
 :AbstractFloat
 :AbstractString
 :Any
 :Bool
 :BoundsError
 :Char
 :CodeInfo
 :Core
 :DataType
 :DenseArray
 :DirectIndexString
 :DivideError
 :DomainError
 :Exception
 :Expr
 :Float16
 ⋮
 :VecElement
 :Void
 :WeakRef
 :applicable
 :eval
 :fieldtype
 :getfield
 :invoke
 :isa
 :isdefined
 :issubtype
 :nfields
 :nothing
 :setfield!
 :throw
 :tuple
 :typeassert
 :typeof
```

**Comment by JeffBezanson** [comment]
The last argument to `names` only includes things explicitly imported (with `import`). `names` never returns something just because the module uses a certain name from somewhere.

**Comment by StefanKarpinski** [comment]
I have to say that the current situation where if the user wants to know what they will have available to them by default, they have to understand the difference between `Core` and `Base` and look in both of them seems like a somewhat unfortunately leaky abstraction.

**Comment by StefanKarpinski** [comment]
In other words, it seems to me that `Base` should re-export everything exported by `Core` so that users don't need to ever think about `Core` or the difference between functions that come from it and those that are part of the standard library – there's been a fair amount of shifting over time.

**Comment by JeffBezanson** [comment]
In what situation do you have to think about `Core`?

`Core` is very tiny and changes glacially. I don't think there has been much shifting between there and Base. Core also exports mostly types and very few functions.

I believe the solution to this is tab completion, help, and other tooling. Those will scale to having many default packages available.

**Comment by tkelman** [comment]
There are a few deprecated command line flags that I think were done before 0.5 and should be deleted now.

**Comment by StefanKarpinski** [comment]
If you think that tab completion is the answer, then we're talking about different things. What is the simplest _programmatic_ way to produce all of the names that are available to you in a new module?
What exactly is the drawback of Base re-exporting everything that Core exports?

**Comment by StefanKarpinski** [comment]
> There are a few deprecated command line flags

That can be done in a separate PR, this should be safe to merge now. Any objections?

File: base/show.jl [hunk 4]
```diff
@@ -82,17 +82,10 @@ haskey(io::IOContext, key) = haskey(io.dict, key)
...
 function get(io::IOContext, key, default)
```
Comment (Sacha0) [inline base/show.jl]: Perhaps reduce these to one-liners?
```julia
get(io::IOContext, key, default) = get(io.dict, key, default)
get(io::IO, key, default) = default
```

**Comment by JeffBezanson** [comment]
I guess it is pretty harmless, but I still don't really understand the use case for it. For programmatic access to module/name info, we typically want information about exactly where things come from. From there, we can recursively look at what a module is `using` to see what else is available within it. We could add a function, or an option to `names`, to do this recursion. I prefer the generality of that to manually combining the names of two specific modules.

**Comment by StefanKarpinski** [comment]
The specific use case was that I was looking for what had disappeared from Base when I removed the deprecations. It looked like String had, which is bogus since it was just no longer exported from Base and is only exported from Core. However, it's entirely plausible that code in the wild is looking to see if `:String in names(Base)` which has now changed, inadvertently. Adding an option to `names` would be fine, although I really think those options need to be keywords.

**Comment by JeffBezanson** [comment]
Agreed they should be keywords. We could even make that change now I'd say.

Part of this is due to `@deprecate`'s insertion of `export` statements. Maybe we should change it to only export things that were previously exported.

**Comment by tkelman** [comment]
I think simon added an option so you can tell deprecate whether or not to export the old name

File: src/jloptions.c [hunk 5]
```diff
@@ -303,10 +293,6 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
         case 'E': // print
             jl_options.print = strdup(optarg);
             break;
-        case 'P': // post-boot
-            jl_printf(JL_STDOUT, "WARNING: julia -P/--post-boot option is deprecated, use -i -e instead.\n");
-            jl_options.postboot = strdup(optarg);
```
Comment (tkelman) [inline src/jloptions.c]: looks like we should delete this field now

File: base/show.jl (hunk 4 above)
Comment (tkelman) [inline base/show.jl]: that would make sense now, yeah - wasn't worth going through the ci queue again for though

**Pull request merged**